nestkernel/module_manager.cpp

```diff
@@ -0,0 +1,148 @@
+/*
+ *  module_manager.cpp
+ *
+ *  This file is part of NEST.
+ *
+ *  Copyright (C) 2004 The NEST Initiative
+ *
+ *  NEST is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  NEST is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with NEST.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "module_manager.h"
+
+#ifdef HAVE_LIBLTDL
+
+// Includes from libnestutil:
+#include "logging.h"
+
+// Includes from nestkernel:
+#include "kernel_manager.h"
+#include "nest_extension_interface.h"
+
+// Includes from sli:
+#include "arraydatum.h"
+
+namespace nest
+{
+
+ModuleManager::ModuleManager()
+  : modules_()
+{
+  lt_dlinit();
+}
+
+ModuleManager::~ModuleManager()
+{
+  finalize( /* reset_kernel = */ true );
+  lt_dlexit();
+}
+
+void
+ModuleManager::initialize( const bool )
+{
+}
+
+void
+ModuleManager::finalize( const bool )
+{
+  // unload all loaded modules
+  for ( const auto& [ name, handle ] : modules_ )
+  {
+    lt_dlclose( handle );
+  }
+  modules_.clear();
+}
+
+void
+ModuleManager::get_status( DictionaryDatum& d )
+{
+  ArrayDatum loaded;
+  for ( const auto& [ name, handle ] : modules_ )
+  {
+    loaded.push_back( new LiteralDatum( name ) );
+  }
+  ( *d )[ names::modules ] = loaded;
+}
+
+void
+ModuleManager::set_status( const DictionaryDatum& d )
+{
+}
+
+void
+ModuleManager::install( const std::string& name )
+{
+  if ( name.empty() )
+  {
+    throw DynamicModuleManagementError( "Module name must not be empty." );
+  }
+
+  if ( modules_.find( name ) != modules_.end() )
+  {
+    throw DynamicModuleManagementError( "Module '" + name + "' is loaded already." );
```

Comment:
You are right, reloading a module without `ResetKernel()` can lead to chaos if there is a network created with an older version of the module. 

> Thinking of it, I wonder if we should not prohibit module loading after any network components have been created and also after Simulate() has been called (in case someone calls it on an empty network).

Shouldn't the call to `Install()` be prohibited if there are some network components already created?

Reply:
@heplesser ping!

Reply:
@clinssen Actually, I am not sure what state one would end up in if we load the same dynamic module twice without unloading it first. The new version of the same name might even contain different code. It seems much safer to prohibit this.

@pnbabu I have now prohibited Install if nodes exist.

The `name`  in the checks here is actually the file name. But I agree that the `name()` field is a bit double up, so we could leave it out for now and then introduce metadata when we know what we want.

Reply:
I'm fine with disallowing loading a module by the same name twice. In that case, I would suggest to work purely on the basis of the filename and leave out the ``name`` member string in the extension module interface for now.